pkgs/development/interpreters/nextflow/default.nix

```diff
@@ -1,54 +1,96 @@
+  openjdk,
+  gradle,
+  wget,
+  which,
+  gnused,
+  gawk,
+  coreutils,
+  bash,
+  versionCheckHook,
+  nixosTests,
 }:
-
-let
-  nextflow =
-stdenv.mkDerivation rec {
+# https://github.com/NixOS/nixpkgs/pull/339197#pullrequestreview-2306770311
+stdenv.mkDerivation (finalAttrs: rec {
   pname = "nextflow";
-  version = "22.10.6";
+  # 24.08.0-edge is compatible with Java 21. The current (as of 2024-09-19)
+  # nextflow release (24.04.4) does not yet support java21, but java19. The
+  # latter is not in nixpkgs(-unstable) anymore.
+  version = "24.08.0-edge";
 
-  src = fetchurl {
-    url = "https://github.com/nextflow-io/nextflow/releases/download/v${version}/nextflow-${version}-all";
-    hash = "sha256-zeYsKxWRnzr0W6CD+yjoAXwCN/AbN5P4HhH1oftnrjY=";
+  src = fetchFromGitHub {
+    owner = "nextflow-io";
+    repo = "nextflow";
+    rev = "6e866ae81ff3bf8a9729e9dbaa9dd89afcb81a4b";
+    hash = "sha256-SA27cuP3iO5kD6u0uTeEaydyqbyJzOkVtPrb++m3Tv0=";
   };
 
   nativeBuildInputs = [
     makeWrapper
-    openjdk17
-    wget
-    which
-    gnused
-    gawk
-    coreutils
+    gradle
   ];
 
-  dontUnpack = true;
+  postPatch = ''
+    # Nextflow invokes the constant "/bin/bash" (not as a shebang) at
+    # several locations so we fix that globally. However, when running inside
+    # a container, we actually *want* "/bin/bash". Thus the global fix needs
+    # to be reverted for this specific use case.
+    substituteInPlace modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy \
+      --replace-fail "['/bin/bash'," "['${bash}/bin/bash'," \
+      --replace-fail "if( containerBuilder ) {" "if( containerBuilder ) {
+                launcher = launcher.replaceFirst(\"/nix/store/.*/bin/bash\", \"/bin/bash\")"
+  '';
+
+  mitmCache = gradle.fetchDeps {
+    inherit (finalAttrs) pname;
+    data = ./deps.json;
+  };
+  __darwinAllowLocalNetworking = true;
+
+  # During the build, some additional dependencies are downloaded ("detached
+  # configuration"). We thus need to run a full build on instead of the default
+  # one.
+  # See https://github.com/NixOS/nixpkgs/pull/339197#discussion_r1747749061
+  gradleUpdateTask = "pack";
+  # See Makefile (`make pack`)
+  preBuild = ''
+    export BUILD_PACK=1
+  '';
+  gradleBuildTask = "pack";
 
   installPhase = ''
     runHook preInstall
 
     mkdir -p $out/bin
-    install -Dm755 $src $out/bin/nextflow
+    install -Dm755 build/releases/nextflow-${version}-dist $out/bin/nextflow
 
     runHook postInstall
   '';
 
   postFixup = ''
     wrapProgram $out/bin/nextflow \
-      --prefix PATH : ${lib.makeBinPath nativeBuildInputs} \
-      --set JAVA_HOME ${openjdk17.home}
+      --prefix PATH : ${
+        lib.makeBinPath [
+          coreutils
+          gawk
+          gnused
+          wget
+          which
+        ]
+      } \
+      --set JAVA_HOME ${openjdk.home}
   '';
 
+  # TODO: Set NXF_HOME=$TMPDIR or similar.
+  nativeInstallCheckInputs = [ versionCheckHook ];
```

Comment:
> I don't see a mean to use the hook right now. Let me know what you think.

Yes so it seems.. I wonder whether this is maybe case is not that extreme, and we should support adding arbitrary arguments to the `env` command. For now, I'll leave it as it is, and we can go with `passthru.tests` on this one.

Reply:
Yes, I also wondered whether this package justifies adapting `versionCheckHook`. This is beyond my exptertise.

I have added & used `testers.testVersion` again.